### PR TITLE
Improve handling of bad records

### PR DESCRIPTION
**Pull request by whyrusleeping**
And some other misc improvements.

### CODE REVIEW

File: repomgr/repomgr.go [hunk 1]
```diff
@@ -885,28 +885,30 @@ func processOp(ctx context.Context, bs blockstore.Blockstore, op *mst.DiffOp) (*
...
+		outop := &RepoOp{
 			Kind:       kind,
 			Collection: parts[0],
 			Rkey:       parts[1],
 			RecCid:     &op.NewCid,
-			Record:     rec,
-		}, nil
+		}
+
+		rec, err := lexutil.CborDecodeValue(blk.RawData())
```
Comment (whyrusleeping) [inline repomgr/repomgr.go:900]: I guess we could move the record decoding up outside of the repomgr, and just include the bytes in the RepoOp from here

File: events/repostream.go [hunk 2]
```diff
@@ -36,20 +36,23 @@ func ConsumeRepoStreamLite(ctx context.Context, con *websocket.Conn, cb LiteStre
 					if err != nil {
 						e := fmt.Errorf("getting record %s (%s) within seq %d for %s: %w", op.Path, *op.Cid, evt.Seq, evt.Repo, err)
 						log.Error(e)
-						return nil
+						continue
 					}
 
 					if lexutil.LexLink(rc) != *op.Cid {
+						// TODO: do we even error here?
```
Comment (bnewbold) [inline events/repostream.go:43]: I would say we should error out here. If the ops metadata is inconsistent with repo slice, that is pretty bad and we should error out and not partially process.

Maybe even mark repo as "dirty"? Something to think about this week.

File: indexer/crawler.go [hunk 3]
```diff
@@ -20,22 +21,29 @@ type CrawlDispatcher struct {
...
+		concurrency: concurrency,
+	}, nil
 }
 
 func (c *CrawlDispatcher) Run() {
 	go c.mainLoop()
 
-	for i := 0; i < 3; i++ {
+	for i := 0; i < c.concurrency; i++ {
 		go c.fetchWorker()
```
Comment (bnewbold) [inline indexer/crawler.go:47]: Will probably want to make this runtime configurable (not just hardcode 10 as below), but we can do that later.

A TODO for me this week will be reviewing shutdown behavior. Not change in behavior here though.

File: repomgr/repomgr.go [hunk 4]
```diff
@@ -885,28 +885,30 @@ func processOp(ctx context.Context, bs blockstore.Blockstore, op *mst.DiffOp) (*
 			return nil, err
 		}
 
-		rec, err := lexutil.CborDecodeValue(blk.RawData())
```
Comment (bnewbold) [inline repomgr/repomgr.go:888]: Semantically, I think we should (must?) validate that the record as acceptable here, but it should be lexicon-agnostic. We probably don't have a helper for that yet? What I mean is things like:

- actually decodes as CBOR at all
- meets additional DAG-CBOR requirements (eg, no tag types other than 42 "link")
- meets any Lexicon requirements, eg no weird CID types (raw-inline data CID)

Basically BGS needs to be conservative about low-level details (below app schema) and permissive about higher-level stuff (above app pschema).

In this PR, would be good to at least do CBOR decode, if we have a generic way to do that. Maybe call `CborTypeExtract`? all records should (must) have a `$type`.

Comment (bnewbold) [reply repomgr/repomgr.go:888]: Oh, now I see that you moved the `CborDecodeValue` below

File: repomgr/repomgr.go (hunk 1 above)
Comment (bnewbold) [reply repomgr/repomgr.go:900]: That sounds roughly correct. Let's have a quick discussion later today about what `repomgr` is responsible for.

**Review by bnewbold — APPROVED** [review]
Mostly left forward-looking comments, which we can follow-up on later this week.

Overall LGTM to deploy in current setup (for search)

File: events/repostream.go (hunk 2 above)
Comment (whyrusleeping) [reply events/repostream.go:43]: yeah, youre probably right

**Pull request merged**